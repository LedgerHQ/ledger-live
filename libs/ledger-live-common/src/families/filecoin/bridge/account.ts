import {
  AmountRequired,
  FeeNotLoaded,
  InvalidAddress,
  NotEnoughBalance,
  RecipientRequired,
} from "@ledgerhq/errors";
import { BigNumber } from "bignumber.js";
import Fil from "@zondax/ledger-filecoin";
import { Observable } from "rxjs";

import { makeAccountBridgeReceive, makeSync } from "../../../bridge/jsHelpers";
import {
  Account,
  AccountBridge,
  AccountLike,
  BroadcastFnSignature,
  Operation,
  SignOperationEvent,
  SignOperationFnSignature,
} from "@ledgerhq/types-live";
import { Transaction, TransactionStatus } from "../types";
import { getAccountShape, getAddress, getTxToBroadcast } from "./utils/utils";
import { broadcastTx, fetchBalances, fetchEstimatedFees } from "./utils/api";
import { getMainAccount } from "../../../account";
import { close } from "../../../hw";
import { toCBOR } from "./utils/serializer";
import { calculateEstimatedFees, getPath, isError } from "../utils";
import { log } from "@ledgerhq/logs";
import { getAddressRaw, validateAddress } from "./utils/addresses";
import { encodeOperationId, patchOperationWithHash } from "../../../operation";
import { withDevice } from "../../../hw/deviceAccess";

const receive = makeAccountBridgeReceive();

const createTransaction = (): Transaction => {
  // log("debug", "[createTransaction] creating base tx");

  return {
    family: "filecoin",
    amount: new BigNumber(0),
    method: 0,
    version: 0,
    nonce: 0,
    gasLimit: new BigNumber(0),
    gasFeeCap: new BigNumber(0),
    gasPremium: new BigNumber(0),
    recipient: "",
    useAllAmount: false,
  };
};

const updateTransaction = (t: Transaction, patch: Transaction): Transaction => {
  // log("debug", "[updateTransaction] patching tx");

  return { ...t, ...patch };
};

const getTransactionStatus = async (
  a: Account,
  t: Transaction
): Promise<TransactionStatus> => {
  // log("debug", "[getTransactionStatus] start fn");

  const errors: TransactionStatus["errors"] = {};
  const warnings: TransactionStatus["warnings"] = {};

  const { balance } = a;
  const { address } = getAddress(a);
  const { recipient, useAllAmount, gasPremium, gasFeeCap, gasLimit } = t;
  let { amount } = t;

  if (!recipient) errors.recipient = new RecipientRequired();
  else if (!validateAddress(recipient).isValid)
    errors.recipient = new InvalidAddress();
  else if (!validateAddress(address).isValid)
    errors.sender = new InvalidAddress();

  if (gasFeeCap.eq(0) || gasPremium.eq(0) || gasLimit.eq(0))
    errors.gas = new FeeNotLoaded();

  // This is the worst case scenario (the tx won't cost more than this value)
  const estimatedFees = calculateEstimatedFees(gasFeeCap, gasLimit);

  let totalSpent;
  if (useAllAmount) {
    totalSpent = a.spendableBalance;
    amount = totalSpent.minus(estimatedFees);
    if (amount.lte(0) || totalSpent.gt(balance)) {
      errors.amount = new NotEnoughBalance();
    }
  } else {
    totalSpent = amount.plus(estimatedFees);
    if (amount.eq(0)) {
      errors.amount = new AmountRequired();
    } else if (totalSpent.gt(a.spendableBalance))
      errors.amount = new NotEnoughBalance();
  }

  // log("debug", "[getTransactionStatus] finish fn");

  return {
    errors,
    warnings,
    estimatedFees,
    amount,
    totalSpent,
  };
};

const estimateMaxSpendable = async ({
  account,
  parentAccount,
  transaction,
}: {
  account: AccountLike;
  parentAccount?: Account | null | undefined;
  transaction?: Transaction | null | undefined;
}): Promise<BigNumber> => {
  // log("debug", "[estimateMaxSpendable] start fn");

  const a = getMainAccount(account, parentAccount);
  const { address } = getAddress(a);

  const recipient = transaction?.recipient;

  if (!validateAddress(address).isValid) throw new InvalidAddress();
  if (recipient && !validateAddress(recipient).isValid)
    throw new InvalidAddress();

  const balances = await fetchBalances(address);
  const balance = new BigNumber(balances.spendable_balance);

  if (balance.eq(0)) return balance;

  const amount = transaction?.amount;

  const result = await fetchEstimatedFees({ to: recipient, from: address });
  const gasFeeCap = new BigNumber(result.gas_fee_cap);
  const gasLimit = new BigNumber(result.gas_limit);
  const estimatedFees = calculateEstimatedFees(gasFeeCap, gasLimit);

  if (balance.lte(estimatedFees)) return new BigNumber(0);

  balance.minus(estimatedFees);
  if (amount) balance.minus(amount);

  // log("debug", "[estimateMaxSpendable] finish fn");

  return balance;
};

const prepareTransaction = async (
  a: Account,
  t: Transaction
): Promise<Transaction> => {
  // log("debug", "[prepareTransaction] start fn");

  const { address } = getAddress(a);
  const { recipient } = t;

  if (recipient && address) {
    // log("debug", "[prepareTransaction] fetching estimated fees");

    if (
      validateAddress(recipient).isValid &&
      validateAddress(address).isValid
    ) {
      const result = await fetchEstimatedFees({ to: recipient, from: address });
      t.gasFeeCap = new BigNumber(result.gas_fee_cap);
      t.gasPremium = new BigNumber(result.gas_premium);
      t.gasLimit = new BigNumber(result.gas_limit);
      t.nonce = result.nonce;
    }
  }

  // log("debug", "[prepareTransaction] finish fn");

  return t;
};

const sync = makeSync({ getAccountShape });

const broadcast: BroadcastFnSignature = async ({
  signedOperation: { operation, signature },
}) => {
  // log("debug", "[broadcast] start fn");

  const tx = getTxToBroadcast(operation, signature);

  const resp = await broadcastTx(tx);
  const { hash } = resp;

  const result = patchOperationWithHash(operation, hash);

  // log("debug", "[broadcast] finish fn");

  return result;
};

const signOperation: SignOperationFnSignature<Transaction> = ({
  account,
  deviceId,
  transaction,
}): Observable<SignOperationEvent> =>
  withDevice(deviceId)(
    (transport) =>
      new Observable((o) => {
        async function main() {
          // log("debug", "[signOperation] start fn");

          const {
            recipient,
            method,
            version,
            nonce,
            gasFeeCap,
            gasLimit,
            gasPremium,
            useAllAmount,
          } = transaction;
          let { amount } = transaction;
          const { id: accountId, balance } = account;
          const { address, derivationPath } = getAddress(account);

          if (!gasFeeCap.gt(0) || !gasLimit.gt(0)) {
            log(
              "debug",
              `signOperation missingData --> gasFeeCap=${gasFeeCap} gasLimit=${gasLimit}`
            );
            throw new FeeNotLoaded();
          }

          const filecoin = new Fil(transport);

          try {
            o.next({
              type: "device-signature-requested",
            });

            const fee = calculateEstimatedFees(gasFeeCap, gasLimit);
            if (useAllAmount) amount = balance.minus(fee);

            transaction = { ...transaction, amount };

            // Serialize tx
            const serializedTx = toCBOR(
              getAddressRaw(address),
              getAddressRaw(recipient),
              transaction
            );

            log(
              "debug",
              `[signOperation] serialized CBOR tx: [${serializedTx.toString(
                "hex"
              )}]`
            );

            // Sign by device
            const result = await filecoin.sign(
              getPath(derivationPath),
              serializedTx
            );
            isError(result);

            o.next({
              type: "device-signature-granted",
            });

            // resolved at broadcast time
            const txHash = "";

            // build signature on the correct format
            const signature = `${result.signature_compact.toString("base64")}`;

            const operation: Operation = {
              id: encodeOperationId(accountId, txHash, "OUT"),
              hash: txHash,
              type: "OUT",
              senders: [address],
              recipients: [recipient],
              accountId,
              value: amount.plus(fee),
              fee,
              blockHash: null,
              blockHeight: null,
              date: new Date(),
              extra: {
                gasLimit,
                gasFeeCap,
                gasPremium,
                method,
                version,
                nonce,
                signatureType: 1,
              },
            };

            o.next({
              type: "signed",
              signedOperation: {
                operation,
                signature,
                expirationDate: null,
              },
            });
          } finally {
            close(transport, deviceId);

            // log("debug", "[signOperation] finish fn");
          }
        }

        main().then(
          () => o.complete(),
          (e) => o.error(e)
        );
      })
  );

export const accountBridge: AccountBridge<Transaction> = {
  createTransaction,
  updateTransaction,
  prepareTransaction,
  getTransactionStatus,
  estimateMaxSpendable,
  sync,
  receive,
  broadcast,
  signOperation,
};
