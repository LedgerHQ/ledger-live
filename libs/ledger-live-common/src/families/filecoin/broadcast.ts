import { AccountBridge } from "@ledgerhq/types-live";
import { patchOperationWithHash } from "@ledgerhq/coin-framework/operation";
import { getTxToBroadcast } from "./bridge/utils/utils";
import { broadcastTx } from "./bridge/utils/api";
import { Transaction } from "./types";

export const broadcast: AccountBridge<Transaction>["broadcast"] = async ({ signedOperation }) => {
  // log("debug", "[broadcast] start fn");
  const { operation, signature, rawData } = signedOperation;
  const tx = getTxToBroadcast(operation, signature, rawData!);

  const resp = await broadcastTx(tx);
  const { hash } = resp;

  const result = patchOperationWithHash(signedOperation.operation, hash);

  // log("debug", "[broadcast] finish fn");

  return result;
};
