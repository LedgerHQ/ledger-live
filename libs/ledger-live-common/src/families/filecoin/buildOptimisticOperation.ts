import { Account, Operation, OperationType } from "@ledgerhq/types-live";
import { encodeOperationId } from "@ledgerhq/coin-framework/operation";
import { Transaction } from "./types";
import { toCBORResponse } from "./bridge/utils/serializer";
import { calculateEstimatedFees } from "./utils";
import { convertAddressFilToEthAsync } from "./bridge/utils/addresses";
import { getSubAccount } from "./bridge/utils/utils";

export const buildOptimisticOperation = async (
  account: Account,
  transaction: Transaction,
  serializationRes: toCBORResponse,
  operationType: OperationType = "OUT",
): Promise<Operation> => {
  const type = operationType;
  const subAccount = getSubAccount(account, transaction);
  // resolved at broadcast time
  const txHash = "";
  const { gasFeeCap, gasLimit } = transaction;
  const { parsedSender, recipientToBroadcast, amountToBroadcast: finalAmount } = serializationRes;
  const fee = calculateEstimatedFees(gasFeeCap, gasLimit);

  let operation: Operation;
  if (subAccount) {
    const senderEthAddr = await convertAddressFilToEthAsync(parsedSender);
    const recipientEthAddr = await convertAddressFilToEthAsync(transaction.recipient);
    operation = {
      id: encodeOperationId(subAccount.id, txHash, "OUT"),
      hash: txHash,
      type,
      value: finalAmount,
      fee,
      blockHeight: null,
      blockHash: null,
      accountId: subAccount.id,
      senders: [senderEthAddr],
      recipients: [recipientEthAddr],
      date: new Date(),
      extra: {},
    };
  } else {
    operation = {
      id: encodeOperationId(account.id, txHash, "OUT"),
      hash: txHash,
      type,
      senders: [parsedSender],
      recipients: [recipientToBroadcast],
      accountId: account.id,
      value: finalAmount.plus(fee),
      fee,
      blockHash: null,
      blockHeight: null,
      date: new Date(),
      extra: {},
    };
  }

  return operation;
};
