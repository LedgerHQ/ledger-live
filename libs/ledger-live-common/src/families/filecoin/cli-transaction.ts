import type {
  Account,
  AccountLike,
  AccountLikeArray,
} from "@ledgerhq/types-live";
import invariant from "invariant";
import flatMap from "lodash/flatMap";
import type { Transaction } from "../../generated/types";

const options = [];

function inferAccounts(account: Account): AccountLikeArray {
  invariant(account.currency.family === "filecoin", "filecoin family");

  const accounts: Account[] = [account];
  return accounts;
}

function inferTransactions(
  transactions: Array<{
    account: AccountLike;
    transaction: Transaction;
    mainAccount: Account;
  }>
): Transaction[] {
  return flatMap(transactions, ({ transaction }) => {
    invariant(transaction.family === "filecoin", "filecoin family");

    return {
      ...transaction,
      family: "filecoin",
    } as Transaction;
  });
}

export default {
  options,
  inferAccounts,
  inferTransactions,
};
