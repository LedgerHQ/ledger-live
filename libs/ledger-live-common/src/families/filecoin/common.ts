export const isNoErrorReturnCode = (code: number) => code === 0x9000;

export const getPath = (path: string) => (path && path.substr(0, 2) !== "m/" ? `m/${path}` : path);

export const isError = (r: { return_code: number; error_message: string }) => {
  if (!isNoErrorReturnCode(r.return_code)) throw new Error(`${r.return_code} - ${r.error_message}`);
};
