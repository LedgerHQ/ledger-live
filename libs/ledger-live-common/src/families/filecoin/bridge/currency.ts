import { CurrencyBridge } from "@ledgerhq/types-live";
import { getAccountShape } from "./utils/utils";
import { makeScanAccounts } from "../../../bridge/jsHelpers";

const scanAccounts = makeScanAccounts({ getAccountShape });

export const currencyBridge: CurrencyBridge = {
  preload: () => Promise.resolve({}),
  hydrate: () => {},
  scanAccounts,
};
