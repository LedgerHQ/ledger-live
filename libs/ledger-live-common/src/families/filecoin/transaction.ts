import type { Transaction, TransactionRaw } from "./types";
import {
  formatTransactionStatusCommon as formatTransactionStatus,
  fromTransactionCommonRaw,
  fromTransactionStatusRawCommon as fromTransactionStatusRaw,
  toTransactionCommonRaw,
  toTransactionStatusRawCommon as toTransactionStatusRaw,
} from "../../transaction/common";
import type { Account } from "@ledgerhq/types-live";
import { getAccountUnit } from "../../account";
import { formatCurrencyUnit } from "../../currencies";
import BigNumber from "bignumber.js";

export const formatTransaction = (
  { recipient, useAllAmount, amount }: Transaction,
  account: Account
): string => `
SEND ${
  useAllAmount
    ? "MAX"
    : amount.isZero()
    ? ""
    : " " +
      formatCurrencyUnit(getAccountUnit(account), amount, {
        showCode: true,
        disableRounding: true,
      })
}
TO ${recipient}`;

export const fromTransactionRaw = (tr: TransactionRaw): Transaction => {
  const common = fromTransactionCommonRaw(tr);
  return {
    ...common,
    family: tr.family,
    version: tr.version,
    method: tr.method,
    nonce: tr.nonce,
    amount: new BigNumber(tr.amount),
    gasFeeCap: new BigNumber(tr.gasFeeCap),
    gasLimit: new BigNumber(tr.gasLimit),
    gasPremium: new BigNumber(tr.gasPremium),
  };
};

const toTransactionRaw = (t: Transaction): TransactionRaw => {
  const common = toTransactionCommonRaw(t);

  return {
    ...common,
    method: t.method,
    version: t.version,
    family: t.family,
    nonce: t.nonce,
    amount: t.amount.toFixed(),
    gasFeeCap: t.gasFeeCap.toString(),
    gasLimit: t.gasLimit.toNumber(),
    gasPremium: t.gasPremium.toString(),
  };
};

export default {
  formatTransaction,
  fromTransactionRaw,
  toTransactionRaw,
  fromTransactionStatusRaw,
  toTransactionStatusRaw,
  formatTransactionStatus,
};
