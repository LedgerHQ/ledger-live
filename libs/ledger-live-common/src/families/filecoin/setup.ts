// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import { createBridges } from "@ledgerhq/coin-filecoin/index";
import Transport from "@ledgerhq/hw-transport";
import FilecoinApp from "@zondax/ledger-filecoin";
import filecoinResolver from "@ledgerhq/coin-filecoin/signer/index";
import { signMessage } from "@ledgerhq/coin-filecoin/hw-signMessage";
import type { Account, Bridge } from "@ledgerhq/types-live";
import makeCliTools from "@ledgerhq/coin-filecoin/test/cli";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import { Resolver } from "../../hw/getAddress/types";
import {
  TransactionStatus,
  Transaction,
  FilecoinGetAddrResponse,
  FilecoinSignature,
  FilecoinSigner,
} from "./types";
import { getPath, isError } from "./common";

const createSigner: CreateSigner<FilecoinSigner> = (transport: Transport) => {
  const filecoin = new FilecoinApp(transport);
  return {
    showAddressAndPubKey: async (path: string): Promise<FilecoinGetAddrResponse> => {
      const r = await filecoin.showAddressAndPubKey(getPath(path));
      isError(r);

      return r;
    },
    getAddressAndPubKey: async (path: string): Promise<FilecoinGetAddrResponse> => {
      const r = await filecoin.getAddressAndPubKey(getPath(path));
      isError(r);

      return r;
    },
    sign: async (path: string, message: Uint8Array): Promise<FilecoinSignature> => {
      const r = await filecoin.sign(getPath(path), message);
      isError(r);

      return r;
    },
  };
};

const bridge: Bridge<Transaction, Account, TransactionStatus> = createBridges(
  executeWithSigner(createSigner),
);

const messageSigner = {
  signMessage,
};

const resolver: Resolver = createResolver(createSigner, filecoinResolver);

const cliTools = makeCliTools();

export { bridge, cliTools, messageSigner, resolver };
