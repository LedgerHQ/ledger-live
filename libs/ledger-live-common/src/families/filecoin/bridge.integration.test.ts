import "../../__tests__/test-helpers/setup";
import { testBridge } from "../../__tests__/test-helpers/bridge";
import { getAccountBridge } from "../../bridge";
import {
  decodeAccountId,
  encodeAccountId,
  fromAccountRaw,
} from "../../account";
import { BigNumber } from "bignumber.js";
import {
  AmountRequired,
  InvalidAddress,
  NotEnoughBalance,
} from "@ledgerhq/errors";

import type { DatasetTest, CurrenciesData } from "@ledgerhq/types-live";
import type { Transaction } from "./types";
import { fromTransactionRaw } from "../filecoin/transaction";

const SEED_IDENTIFIER = "f1p74d4mlmeyc4agflhjqsnvoyzyfdai7fmkyso2a";
const ACCOUNT_1 = "f1p74d4mlmeyc4agflhjqsnvoyzyfdai7fmkyso2a";

const filecoin: CurrenciesData<Transaction> = {
  FIXME_ignoreAccountFields: [],
  scanAccounts: [
    {
      name: "filecoin seed 1",
      apdus: `
      => 0600000000
      <= 0000160500311000049000
      => 06010000142c000080cd010080000000800000000000000000
      <= 04ca7b02cafdf36e8b4caaf530a96b949764af71b956b2a3328b7a10940794c860f574a9199be98bde3c261887fec8e5fd94bc5f104908bf5f992f52ef2a89abb015017ff83e316c2605c018ab3a6126d5d8ce0a3023e529663170373464346d6c6d657963346167666c686a71736e766f797a796664616937666d6b79736f32619000
      => 0600000000
      <= 0000160500311000049000
      => 06010000142c00008001000080000000800000000000000000
      <= 04d493dab3ff63d6f0673454f8f6d6adec08db19e8c0298c65cfcace76a10757774afdee518b15a8b82452531bcb7860eefce02a89ef9051037646f5d4d6bd74171501eac14c6468ac13f66f0e0ee3766472c7e98394ab297431356c6175797a646976716a376d33796f623372786d7a6473793775796866666c777968657572699000
      => 0600000000
      <= 0000160500311000049000
      => 06010000142c00008001000080000000800000000001000000
      <= 040cbb57791d797268e34340565393f8a014ae754026159973fbe46bdcef3b6d28d155f4923256a2beb9bcff3bd39e5066d1c778306f28a2d4faa3a66dbf70453e150148c9b66fef1fce116652099bec6301ced31e56472974316a6465336d33377064376862637a737362676e36797979627a336a7234767368676370696b32619000
      => 0600000000
      <= 0000160500311000049000
      => 06010000142c000080cd010080000000800000000001000000
      <= 0419d91b4f9c3015637f5aa6b601b337f2770b1dcf93ba896d64627cd2aa67493ba924decd3024643478a84e06486b0a6b10d9f7607b6e92b0545bb3c998d4d75d15017ad86da94bab1ae1942e43050edb5f634b2d1ec9296631706c6d67336b6b6c766d6e6f6466626f696d6371357732376d6e66733268776a746173366769619000
      => 0600000000
      <= 0000160500311000049000
      => 06010000142c000080cd010080010000800000000000000000
      <= 0463c18e9a3c08ecffab4fa384f977e5d380d2b252dbbd96875df9cc91c44c81e82639fd659a3eeeaed96901696d8c7832b03befafd0af6ab2bee634ea8141e52a15016f62da10a416f416e80ec32866b91f01f7746af82966316e35726e75656665633332626e32616f796d75676e6f693761683378693278796e766c746e71699000
      `,
    },
  ],
  accounts: [
    {
      FIXME_tests: ["balance is sum of ops"],
      raw: {
        id: `js:2:filecoin:${SEED_IDENTIFIER}:glif`,
        seedIdentifier: SEED_IDENTIFIER,
        name: "Filecoin 1",
        derivationMode: "glif" as const,
        index: 0,
        freshAddress: SEED_IDENTIFIER,
        freshAddressPath: "44'/461'/0'/0/0",
        freshAddresses: [],
        blockHeight: 0,
        operations: [],
        pendingOperations: [],
        currencyId: "filecoin",
        unitMagnitude: 18,
        lastSyncDate: "",
        balance: "1000",
      },
      transactions: [
        {
          name: "Not a valid address",
          transaction: fromTransactionRaw({
            family: "filecoin",
            method: 1,
            version: 1,
            nonce: 100,
            gasFeeCap: "1000",
            gasLimit: 100,
            gasPremium: "200",
            recipient: "novalidaddress",
            amount: "100000000",
          }),
          expectedStatus: {
            errors: {
              recipient: new InvalidAddress(),
            },
            warnings: {},
          },
        },
        {
          name: "Not enough balance",
          transaction: fromTransactionRaw({
            family: "filecoin",
            method: 1,
            version: 1,
            nonce: 100,
            gasFeeCap: "1000",
            gasLimit: 10,
            gasPremium: "10000",
            recipient: ACCOUNT_1,
            amount: "100000000000000000000",
          }),
          expectedStatus: {
            errors: {
              amount: new NotEnoughBalance(),
            },
            warnings: {},
          },
        },
        {
          name: "Amount Required",
          transaction: fromTransactionRaw({
            family: "filecoin",
            method: 1,
            version: 1,
            nonce: 100,
            gasFeeCap: "10",
            gasLimit: 10,
            gasPremium: "10000",
            recipient: ACCOUNT_1,
            amount: "0",
          }),
          expectedStatus: {
            errors: {
              amount: new AmountRequired(),
            },
            warnings: {},
          },
        },
        {
          name: "New account and sufficient amount",
          transaction: fromTransactionRaw({
            family: "filecoin",
            method: 1,
            version: 1,
            nonce: 100,
            gasFeeCap: "1000",
            gasLimit: 10,
            gasPremium: "10000",
            recipient: ACCOUNT_1,
            amount: "1",
          }),
          expectedStatus: {
            amount: new BigNumber("1"),
            errors: {},
            warnings: {},
          },
        },
      ],
    },
  ],
};

const dataset: DatasetTest<Transaction> = {
  implementations: ["js"],
  currencies: {
    filecoin,
  },
};

testBridge(dataset);

describe("estimateMaxSpendable", () => {
  test("it should failed on invalid recipient", async () => {
    const accounts = dataset.currencies["filecoin"].accounts || [];
    const accountData = accounts[0];

    const account = fromAccountRaw({
      ...accountData.raw,
      id: encodeAccountId({
        ...decodeAccountId(accountData.raw.id),
        type: dataset.implementations[0],
      }),
    });

    const accountBridge = getAccountBridge(account);
    const estimate = async () => {
      await accountBridge.estimateMaxSpendable({
        account,
        transaction: { recipient: "notavalidrecipient" },
      });
    };

    await expect(estimate).rejects.toThrowError(new InvalidAddress());
  });
});
