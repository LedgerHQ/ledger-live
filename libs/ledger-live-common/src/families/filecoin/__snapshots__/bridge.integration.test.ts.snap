// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filecoin currency bridge scanAccounts filecoin seed 1 1`] = `
Array [
  Object {
    "balance": "1000000000000000",
    "currencyId": "filecoin",
    "derivationMode": "gliflegacy",
    "freshAddress": "t15lauyzdivqj7m3yob3rxmzdsy7uyhfflwyheuri",
    "freshAddressPath": "44'/1'/0'/0/0",
    "freshAddresses": Array [
      Object {
        "address": "t15lauyzdivqj7m3yob3rxmzdsy7uyhfflwyheuri",
        "derivationPath": "44'/1'/0'/0/0",
      },
    ],
    "id": "js:2:filecoin:t15lauyzdivqj7m3yob3rxmzdsy7uyhfflwyheuri:gliflegacy",
    "index": 0,
    "name": "Filecoin 1",
    "nfts": undefined,
    "operationsCount": 1,
    "pendingOperations": Array [],
    "seedIdentifier": "04ca7b02cafdf36e8b4caaf530a96b949764af71b956b2a3328b7a10940794c860f574a9199be98bde3c261887fec8e5fd94bc5f104908bf5f992f52ef2a89abb0",
    "spendableBalance": "1000000000000000",
    "starred": false,
    "swapHistory": Array [],
    "syncHash": undefined,
    "unitMagnitude": 18,
    "used": true,
  },
  Object {
    "balance": "100000000000000000",
    "currencyId": "filecoin",
    "derivationMode": "",
    "freshAddress": "f1p74d4mlmeyc4agflhjqsnvoyzyfdai7fmkyso2a",
    "freshAddressPath": "44'/461'/0'/0/0",
    "freshAddresses": Array [
      Object {
        "address": "f1p74d4mlmeyc4agflhjqsnvoyzyfdai7fmkyso2a",
        "derivationPath": "44'/461'/0'/0/0",
      },
    ],
    "id": "js:2:filecoin:f1p74d4mlmeyc4agflhjqsnvoyzyfdai7fmkyso2a:",
    "index": 0,
    "name": "Filecoin 1",
    "nfts": undefined,
    "operationsCount": 1,
    "pendingOperations": Array [],
    "seedIdentifier": "04ca7b02cafdf36e8b4caaf530a96b949764af71b956b2a3328b7a10940794c860f574a9199be98bde3c261887fec8e5fd94bc5f104908bf5f992f52ef2a89abb0",
    "spendableBalance": "100000000000000000",
    "starred": false,
    "swapHistory": Array [],
    "syncHash": undefined,
    "unitMagnitude": 18,
    "used": true,
  },
  Object {
    "balance": "0",
    "currencyId": "filecoin",
    "derivationMode": "",
    "freshAddress": "f1n5rnuefec32bn2aoymugnoi7ah3xi2xynvltnqi",
    "freshAddressPath": "44'/461'/1'/0/0",
    "freshAddresses": Array [
      Object {
        "address": "f1n5rnuefec32bn2aoymugnoi7ah3xi2xynvltnqi",
        "derivationPath": "44'/461'/1'/0/0",
      },
    ],
    "id": "js:2:filecoin:f1n5rnuefec32bn2aoymugnoi7ah3xi2xynvltnqi:",
    "index": 1,
    "name": "Filecoin 2",
    "nfts": undefined,
    "operationsCount": 0,
    "pendingOperations": Array [],
    "seedIdentifier": "04ca7b02cafdf36e8b4caaf530a96b949764af71b956b2a3328b7a10940794c860f574a9199be98bde3c261887fec8e5fd94bc5f104908bf5f992f52ef2a89abb0",
    "spendableBalance": "0",
    "starred": false,
    "swapHistory": Array [],
    "syncHash": undefined,
    "unitMagnitude": 18,
    "used": false,
  },
]
`;

exports[`filecoin currency bridge scanAccounts filecoin seed 1 2`] = `
Array [
  Array [
    Object {
      "accountId": "js:2:filecoin:t15lauyzdivqj7m3yob3rxmzdsy7uyhfflwyheuri:gliflegacy",
      "blockHash": null,
      "blockHeight": 1899682,
      "contract": undefined,
      "extra": Object {},
      "fee": "0",
      "hash": "bafy2bzaceawypi647m4rz5czma3xm7syif67bfzce5q2w2tifd5jrvqeaowok",
      "id": "js:2:filecoin:t15lauyzdivqj7m3yob3rxmzdsy7uyhfflwyheuri:gliflegacy-bafy2bzaceawypi647m4rz5czma3xm7syif67bfzce5q2w2tifd5jrvqeaowok-IN",
      "operator": undefined,
      "recipients": Array [
        "t15lauyzdivqj7m3yob3rxmzdsy7uyhfflwyheuri",
      ],
      "senders": Array [
        "f1ov6d42tujoyexkbdh34oik2vhe5unqo2a5ocqoq",
      ],
      "standard": undefined,
      "tokenId": undefined,
      "type": "IN",
      "value": "1000000000000000",
    },
  ],
  Array [
    Object {
      "accountId": "js:2:filecoin:f1p74d4mlmeyc4agflhjqsnvoyzyfdai7fmkyso2a:",
      "blockHash": null,
      "blockHeight": 1802367,
      "contract": undefined,
      "extra": Object {},
      "fee": "0",
      "hash": "bafy2bzaceaztepnzdzbul7nmkef6blk5x2luvbvangrjal25k6xi5qob36k32",
      "id": "js:2:filecoin:f1p74d4mlmeyc4agflhjqsnvoyzyfdai7fmkyso2a:-bafy2bzaceaztepnzdzbul7nmkef6blk5x2luvbvangrjal25k6xi5qob36k32-IN",
      "operator": undefined,
      "recipients": Array [
        "f1p74d4mlmeyc4agflhjqsnvoyzyfdai7fmkyso2a",
      ],
      "senders": Array [
        "f1ov6d42tujoyexkbdh34oik2vhe5unqo2a5ocqoq",
      ],
      "standard": undefined,
      "tokenId": undefined,
      "type": "IN",
      "value": "100000000000000000",
    },
  ],
  Array [],
]
`;
